### PR TITLE
feat: better merge algorithm

### PR DESCRIPTION
**Pull request by mitchellwrosen**
This PR contains the MVP / minimal thing / version 0 of the new merge algorithm.

See [merge.output.md](https://github.com/unisonweb/unison/blob/topic/merge4/unison-src/transcripts/merge.output.md) for an overview of its behavior.

In summary, we diff namespaces based on "syntactic hashes" in which a pretty-print environment renders references before hashing. This is the mechanism by which we distinguish "user changes" from auto-propagated changes.

There are a number of different preconditions that each namespace (and their LCA) must satisfy before we can attempt a merge, such as having no conflicted names, and having exactly one properly-placed name for each constructor of a particular naming of a data declaration. See the above transcript for more details.

A couple outstanding things:
- The old `merge` still exists (hidden) as `merge.old`. The `pull` command still performs a `merge.old` after downloading entities, not a `merge`.
- There's not yet a replacement for `merge.preview` nor `merge.squash`. Those still exist as `merge.old.preview` and `merge.old.squash`.
- We still want to #4928

And a note about the code quality:
- Whoops, a lot of stuff is still badly organized and badly documented. Some helper functions and variables are probably still called `honk`, `bonk`, and perhaps `oink`. However, none of it was feeling like it should block the review of this PR. I plan to continue to refine things in the upcoming days.

### CODE REVIEW

**Comment by pchiusano** [comment]
ITS HAPPENING!!! 🚀🎉

File: lib/unison-prelude/src/Unison/Util/Map.hs [hunk 1]
```diff
@@ -41,6 +61,17 @@ swap :: (Ord b) => Map a b -> Map b a
 swap =
   Map.foldlWithKey' (\z a b -> Map.insert b a z) mempty
 
+-- | Like 'Map.insert', but returns the value as well.
+insertLookup :: Ord k => k -> v -> Map k v -> (Maybe v, Map k v)
```
Comment (aryairani) [inline lib/unison-prelude/src/Unison/Util/Map.hs]: Does this return the old value or the new value?

File: lib/unison-prelude/src/Unison/Util/Map.hs [hunk 2]
```diff
@@ -51,6 +82,11 @@ upsertF :: (Functor f, Ord k) => (Maybe v -> f v) -> k -> Map k v -> f (Map k v)
 upsertF f =
   Map.alterF (fmap Just . f)
 
+-- | Like 'upsert', but returns the value as well.
+upsertLookup :: Ord k => (Maybe v -> v) -> k -> Map k v -> (Maybe v, Map k v)
```
Comment (aryairani) [inline lib/unison-prelude/src/Unison/Util/Map.hs]: ditto

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs [hunk 3]
```diff
@@ -61,6 +64,16 @@ termNameOrHashOnly ppe r = maybe (HQ.fromReferent r) HQ'.toHQ $ terms ppe r
 typeNameOrHashOnly :: PrettyPrintEnv -> Reference -> HQ.HashQualified Name
 typeNameOrHashOnly ppe r = maybe (HQ.fromReference r) HQ'.toHQ $ types ppe r
 
+-- Like 'termNameOrHashOnly' but returns the fully qualified name
```
Comment (aryairani) [inline parser-typechecker/src/Unison/PrettyPrintEnv.hs:67]: What does `termNameOrHashOnly` do?

File: parser-typechecker/src/Unison/Syntax/FileParser.hs [hunk 4]
```diff
@@ -114,55 +116,74 @@ file = do
...
-            (terms <> join accessors)
-            (List.multimap watches)
-    validateUnisonFile uf
-    pure uf
+    -- let uf =
+    --       UnisonFileId
+    --         (UF.datasId env)
+    --         (UF.effectsId env)
+    --         (List.foldl' (\acc (a, b, c) -> Map.insert a (b, c) acc) Map.empty (terms <> join accessors))
+    --         (List.multimap watches)
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Syntax/FileParser.hs]: keep or nah?

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 5]
```diff
@@ -1195,6 +1197,8 @@ inputDescription input =
       src <- either (pure . Text.pack . show) brp src0
       dest <- brp dest0
       pure ("fork " <> src <> " " <> dest)
+    MergeI _tgt -> do
+      pure "todo"
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs]: todo

File: unison-merge/src/Unison/Merge.hs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+
+module Unison.Merge
+  ( 
+  )
+where
+
```
Comment (aryairani) [inline unison-merge/src/Unison/Merge.hs]: Do we want to keep this module?

File: unison-merge/src/Unison/Merge/TwoDiffOps.hs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  | TwoDiffOps'UpdateDelete !(Updated a) -- old was deleted by Bob
+  | TwoDiffOps'UpdateUpdate !a !(TwoWay a) -- old, new
+
+-- | Combine two aligned @DiffOp@ into one @TwoDiffOps@.
+make :: These (DiffOp a) (DiffOp a) -> TwoDiffOps a
+make = \case
+  This alice -> one Alice alice
+  That bob -> one Bob bob
+  These (DiffOp'Add alice) (DiffOp'Add bob) -> TwoDiffOps'AddAdd TwoWay {alice, bob}
+  These (DiffOp'Delete alice) (DiffOp'Delete _bob) -> TwoDiffOps'DeleteDelete alice
```
Comment (aryairani) [inline unison-merge/src/Unison/Merge/TwoDiffOps.hs]:
```suggestion
  These (DiffOp'Delete alice) (DiffOp'Delete _) -> TwoDiffOps'DeleteDelete alice
```

File: unison-src/transcripts/branch-command.output.md [hunk 8]
```diff
@@ -30,85 +30,85 @@ foo/main> branch topic1
...
   Done. I've created the topic1 branch based off of main.
   
-  Tip: Use `merge /topic1 /main` to merge your work back into
-       the main branch.
+  Tip: To merge your work back into the main branch, first
+       `switch /main` then `merge /topic1`.
 
 foo/main> branch /topic2
 
   Done. I've created the topic2 branch based off of main.
```
Comment (aryairani) [inline unison-src/transcripts/branch-command.output.md:38]: unrelated issue, but maybe we want to use the slash syntax here too?

File: unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs [hunk 9]
```diff
@@ -264,29 +278,65 @@ typecheckedUnisonFileToBranchUpdates abort getConstructors tuf = do
...
+                  (Decl.declConstructorReferents typeRefId decl)
+           in insertTypeAction : insertTypeConstructorActions
+
+    termAdds :: [(Path, Branch0 m -> Branch0 m)]
+    termAdds =
+      tuf
+        & UF.hashTermsId
+        & Map.toList
+        & mapMaybe \(var, (_, ref, wk, _, _)) -> do
+          guard (WK.watchKindShouldBeStoredInDatabase wk)
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs:309]: What's this guard about?

File: lib/unison-prelude/src/Unison/Util/Map.hs (hunk 1 above)
Comment (mitchellwrosen) [reply lib/unison-prelude/src/Unison/Util/Map.hs]: It returns the old value. Returning the new would be pretty pointless 😛

File: lib/unison-prelude/src/Unison/Util/Map.hs (hunk 2 above)
Comment (mitchellwrosen) [reply lib/unison-prelude/src/Unison/Util/Map.hs]: The old value

File: parser-typechecker/src/Unison/PrettyPrintEnv.hs (hunk 3 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/PrettyPrintEnv.hs:67]: It returns the best suffixified term name out of a pretty print env for a referent

File: parser-typechecker/src/Unison/Syntax/FileParser.hs (hunk 4 above)
Comment (mitchellwrosen) [reply parser-typechecker/src/Unison/Syntax/FileParser.hs]: oops, no, accidentally left in

File: unison-merge/src/Unison/Merge.hs (hunk 6 above)
Comment (mitchellwrosen) [reply unison-merge/src/Unison/Merge.hs]: nope!

File: unison-src/transcripts/branch-command.output.md (hunk 8 above)
Comment (mitchellwrosen) [reply unison-src/transcripts/branch-command.output.md:38]: I don't think so; the slash isn't part of the branch name. Maybe that's also an argument against using it in the "I merged /bob into /alice" output message 🤔

File: unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs (hunk 9 above)
Comment (mitchellwrosen) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput/Update2.hs:309]: It's keeping only these:

```
foo = ...

test> bar = ...
```

but not

```
> hey
```

nor

```
wat> x = ...
```

File: lib/unison-prelude/src/Unison/Util/Map.hs (hunk 1 above)
Comment (aryairani) [inline lib/unison-prelude/src/Unison/Util/Map.hs]:
```suggestion
-- | Like 'Map.insert', but returns the old value as well.
insertLookup :: Ord k => k -> v -> Map k v -> (Maybe v, Map k v)
```

File: lib/unison-prelude/src/Unison/Util/Map.hs (hunk 2 above)
Comment (aryairani) [inline lib/unison-prelude/src/Unison/Util/Map.hs]:
```suggestion
-- | Like 'upsert', but returns the old value as well.
upsertLookup :: Ord k => (Maybe v -> v) -> k -> Map k v -> (Maybe v, Map k v)
```

**Review by aryairani — APPROVED** [review]
I still owe a more detailed review later, but this is at least great and I wanna avoid more merge conflicts

**Comment by pchiusano** [comment]
Yay!!!!!

**Pull request merged**